resolwe/flow/expression_engines/jinja/__init__.py

```diff
@@ -0,0 +1,96 @@
+"""Jinja2-based expression engine."""
+from importlib import import_module
+
+import jinja2
+from django.core.exceptions import ImproperlyConfigured
+
+from resolwe.flow.expression_engines import BaseExpressionEngine, EvaluationError
+
+from .filters import filters as builtin_filters
+
+
+class NestedUndefined(jinja2.Undefined):
+    """An undefined variable that can be nested."""
+
+    def __unicode__(self):
+        """Override."""
+        return u''
+
+    def __str__(self):
+        """Override."""
+        return u''
+
+    def __call__(self, *args, **kwargs):
+        """Override."""
+        return NestedUndefined()
+
+    def __getattr__(self, *args, **kwargs):
+        """Override."""
+        return NestedUndefined()
+
+
+class ExpressionEngine(BaseExpressionEngine):
+    """Jinja2-based expression engine."""
+
+    name = 'jinja'
+
+    def __init__(self, *args, **kwargs):
+        """Construct the expression engine."""
+        super(ExpressionEngine, self).__init__(*args, **kwargs)
+
+        # Initialize the Jinja2 environment.
+        # TODO: Should we use a sandboxed environment?
+        self._environment = jinja2.Environment(
+            undefined=NestedUndefined
+        )
+        # Register built-in filters.
+        self._environment.filters.update(builtin_filters)
+        # Register custom filters.
+        self._register_custom_filters()
+
+    def _register_custom_filters(self):
+        """Register any custom filter modules."""
+        custom_filters = self.settings.get('CUSTOM_FILTERS', [])
+        if not isinstance(custom_filters, list):
+            raise KeyError("`CUSTOM_FILTERS` setting must be a list.")
+
+        for filter_module_name in custom_filters:
+            try:
+                filter_module = import_module('{}'.format(filter_module_name))
```

Comment:
To avoid saying `str(x)` or `unicode(x)`, which could be different among Python 2/3. I've seen this elsewhere in Resolwe so I used a similar pattern here.